components/ImageGallery.js

```diff
@@ -0,0 +1,356 @@
+      // If controls are visible and image is transformed,
+      // We should switch to image preview mode
+      this.setImagePreviewMode();
+    } else if (mode === IMAGE_PREVIEW_MODE) {
+      this.updateImageSwitchingStatus();
+    }
+  }
+
+  onSingleTapConfirmed() {
+    const { mode } = this.state;
+
+    if (mode === IMAGE_PREVIEW_MODE) {
+      // If controls are not visible and user taps on image
+      // We should switch to gallery mode
+      this.setGalleryMode();
+    } else {
+      this.setImagePreviewMode();
+    }
+  }
+
+  renderDescription(description) {
+    const { collapsed } = this.state;
+    const { style } = this.props;
+
+    if (!description) return;
+
+    const descriptionIcon = collapsed ? <Icon name="up-arrow" /> : <Icon name="down-arrow" />;
+
+    const descriptionText = (
+      <Subtitle
+        style={style.description.text}
+        numberOfLines={collapsed ? 2 : null}
+      >
+        {description}
+      </Subtitle>
+    );
+
+    return (
+      <View
+        style={style.description.container}
+        driver={this.timingDriver}
+        animationName="fadeOut"
+      >
+        <TouchableOpacity onPress={collapsed ? this.collapseDescription : this.expandDescription}>
+          {description.length >= DESCRIPTION_LENGTH_TRIM_LIMIT ? descriptionIcon : null}
+        </TouchableOpacity>
+        <ScrollView
+          onScroll={this.onDescriptionScroll}
+          scrollEventThrottle={200}
+          style={style.description.scroll}
+        >
+          {descriptionText}
+        </ScrollView>
+      </View>
+    );
+  }
+
+  renderTitle(title) {
+    const { style } = this.props;
+
+    return (
+      <View
+        style={style.title.container}
+        driver={this.timingDriver}
+        animationName="fadeOut"
+      >
+        <Subtitle style={style.title.text} numberOfLines={2}>{title}</Subtitle>
+      </View>
+    );
+  }
+
+  getImageTransformer(page) {
+    const { data } = this.props;
+    if (page >= 0 && page < data.length) {
+      const ref = this.imageRefs.get(page);
+      if (ref) {
+        return ref.getViewTransformerInstance();
+      }
+    }
+    return null;
+  }
+
+  resetImageTransformer(transformer) {
+    transformer.updateTransform({ scale: 1, translateX: 0, translateY: 0 });
+  }
+
+  resetSurroundingImageTransformations() {
+    const { selectedIndex } = this.state;
+    let transformer = this.getImageTransformer(selectedIndex - 1);
+    if (transformer) {
+      this.resetImageTransformer(transformer);
+    }
+    transformer = this.getImageTransformer(selectedIndex + 1);
+    if (transformer) {
+      this.resetImageTransformer(transformer);
+    }
+  }
+
+  renderPage(page, pageId) {
+    const { style } = this.props;
+    const { imageSwitchingEnabled } = this.state;
+    const image = _.get(page, 'source.uri');
+    const title = _.get(page, 'title');
+    const description = _.get(page, 'description');
+
+    if (!image) {
+      return;
```

Comment:
Return `null`? How does this affect gallery? Should we maybe have fallback image?